application/models/FailedLoginAttempt.php

```diff
@@ -126,4 +137,34 @@ public function addAttempt()
         }
         return true;
     }
+
+    /**
+     * Returns true if the specified IP is whitelisted
+     *
+     * @param string $ip
+     * @param string $attemptType   'login' or 'token'
+     *
+     * @throws InvalidArgumentException if an invalid attempt type is specified
+     * @return boolean
+     */
+    private function isWhitelisted($ip, $attemptType)
```

Comment:
NB: This method does not contain `this`, so maybe it doesn't belong in a class...? But currently, there's no good system to put functions together with a domain entity, or "module".

Reply:
static function ?

Reply:
Static functions cannot be mocked. But, sometimes they don't need to be mocked. :) (When they have no side-effects.)

This functions has one side-effect: It reads from app config. This should be lifted out if you want to unit-test this function.

Reply:
read App->getConfig are side effect ???

Easy test : 
```
App()->setConfig("myconfigtotest","testvalue");
/* My test #1 */
App()->setConfig("myconfigtotest","testvalue2");
/* My test #2 */
```

Reply:
Accessing globals are considered side-effects.

You can do setConfig, but then you have setup/teardown in the tests. Easy to forget, and tests can affect each other.

Reply:
Oh, and it might be hard to run tests in parallel when you rely on global scope like that.